src/tileiterator.jl

```diff
@@ -0,0 +1,224 @@
+
+Decompose `axes` into an iterator `titr` of smaller axes according to `strategy`.
+
+The `strategy` argument controls the details of the tiling. For instance
+if the length of an axis is not divisible by the tile size, what should happen?
+One approach would be to relax the size requirement for the last tile.
+Another possibility to relax the `stride` so that all tiles are of the requested size,
+but tiles may be slightly overlapping.
+These two possibilities are implemented by [`RelaxLastTile`](@ref) and [`RelaxStride`](@ref).
+
+# Examples
+```jldoctest
+julia> using TiledIteration
+
+julia> collect(TileIterator((1:3, 0:5), RelaxLastTile((2, 3))))
+2×2 Array{Tuple{UnitRange{Int64},UnitRange{Int64}},2}:
+ (1:2, 0:2)  (1:2, 3:5)
+ (3:3, 0:2)  (3:3, 3:5)
+
+julia> collect(TileIterator((1:3, 0:5), (2, 3))) # defaults to RelaxLastTile
+2×2 Array{Tuple{UnitRange{Int64},UnitRange{Int64}},2}:
+ (1:2, 0:2)  (1:2, 3:5)
+ (3:3, 0:2)  (3:3, 3:5)
+
+julia> collect(TileIterator((1:3, 0:5), RelaxStride((2, 3))))
+2×2 Array{Tuple{UnitRange{Int64},UnitRange{Int64}},2}:
+ (1:2, 0:2)  (1:2, 3:5)
+ (2:3, 0:2)  (2:3, 3:5)
+```
+"""
+function TileIterator(axes, strategy)
+    covers1d = map(cover1d, axes, split(strategy))
+    return TileIterator(covers1d)
+end
+
+# strategies
+export RelaxStride
+"""
+    RelaxStride(tilesize)
+
+Tiling strategy, that guarantees each tile of size `tilesize`.
+If the needed tiles will slightly overlap, to cover everything.
+
+# Examples
+```jldoctest
+julia> using TiledIteration
+
+julia> collect(TileIterator((1:4,), RelaxStride((2,))))
+2-element Array{Tuple{UnitRange{Int64}},1}:
+ (1:2,)
+ (3:4,)
+
+julia> collect(TileIterator((1:4,), RelaxStride((3,))))
+2-element Array{Tuple{UnitRange{Int64}},1}:
+ (1:3,)
+ (2:4,)
+```
+
+See also [`TileIterator`](@ref).
+"""
+struct RelaxStride{N}
+    tilesize::Dims{N}
+end
+
+export RelaxLastTile
+
+"""
+    RelaxLastTile(tilesize)
+
+Tiling strategy, that premits the size of the last tiles along each dimension to be smaller
+then `tilesize` if needed. All other tiles are of size `tilesize`.
+
+# Examples
+```jldoctest
+julia> using TiledIteration
+
+julia> collect(TileIterator((1:4,), RelaxLastTile((2,))))
+2-element Array{Tuple{UnitRange{Int64}},1}:
+ (1:2,)
+ (3:4,)
+
+julia> collect(TileIterator((1:7,), RelaxLastTile((2,))))
+4-element Array{Tuple{UnitRange{Int64}},1}:
+ (1:2,)
+ (3:4,)
+ (5:6,)
+ (7:7,)
+```
+
+See also [`TileIterator`](@ref).
+"""
+struct RelaxLastTile{N}
+    tilesize::Dims{N}
+end
+
+"""
+    split(strategy)
+
+Split an N dimensional strategy into an NTuple of 1 dimensional strategies.
+"""
+function split end
+
+function split(strategy::RelaxStride)
+    map(strategy.tilesize) do s
+        RelaxStride((s,))
+    end
+end
+
+function split(strategy::RelaxLastTile)
+    map(strategy.tilesize) do s
+        RelaxLastTile((s,))
+    end
+end
+
+function cover1d(ax, strategy::RelaxStride{1})::CoveredRange
```

Comment:
Is the return type assertion doing anything? If so then you have an inference problem here, but I'm not seeing it by eye.

Reply:
No it is just "documentation" should I drop it?

Reply:
Maybe put a `#` in front of it, just to be clear it's not functional.

Reply:
I just removed it.